Fixed #23829 -- Added support for https scheme and site_domain in ping_google

CuriousLearner opened the pull request:
This PR adds optional arguments to `ping_google` to support changing to `https` scheme using `secure` option and overriding the domain of sitemaps using `site_domain`.

https://code.djangoproject.com/ticket/23829

adamchainz commented on docs/ref/contrib/sitemaps.txt:
```diff
@@ -483,11 +483,18 @@ that: :func:`django.contrib.sitemaps.ping_google()`.
+          to pass your site's domain (e.g., ``example.com``)
+
+        * ``-s --is_secure``, A boolean to indicate to use ``https://`` scheme instead of
+          ``http://``. Defaults to ``False``.
```
you need the `versionadded` annotation for the flags added in django 2.2

adamchainz replied:
wait, you've documented the flags under the function documentation. This should really be about `site_domain` not `-d --site_domain`. The flags should be documented in the last section, "Pinging Google via manage.py".

adamchainz commented on django/contrib/sitemaps/__init__.py:
```diff
@@ -47,7 +48,12 @@ def _get_sitemap_full_url(sitemap_url):
     current_site = Site.objects.get_current()
-    return 'http://%s%s' % (current_site.domain, sitemap_url)
+    if not site_domain:
+        site_domain = current_site.domain
```
lines 49 and 50 should be under the `if` too, otherwise `sites` is still being used

CuriousLearner replied:
Yeah, good catch! Thank you :)

adamchainz commented on django/contrib/sitemaps/__init__.py:
```diff
@@ -15,21 +15,22 @@ class SitemapNotFound(Exception):
-        raise ImproperlyConfigured("ping_google requires django.contrib.sites, which isn't installed.")
+def _get_sitemap_full_url(sitemap_url, site_domain=None, is_secure=False):
+    if not site_domain:
+        if not django_apps.is_installed('django.contrib.sites'):
```
combine with above `if` - `if not site_domain and not django_apps.is_installed...`. save an indentation level.

CuriousLearner commented:
@adamchainz I've addressed your reviews. Can you please do another pass?

adamchainz commented on django/contrib/sitemaps/__init__.py:
```diff
@@ -15,20 +15,20 @@ class SitemapNotFound(Exception):
-    if not django_apps.is_installed('django.contrib.sites'):
+def _get_sitemap_full_url(sitemap_url, site_domain=None, is_secure=False):
+    if not site_domain and not django_apps.is_installed('django.contrib.sites'):
         raise ImproperlyConfigured("ping_google requires django.contrib.sites, which isn't installed.")
```
actually this message could be made more accurate and say "ping_google without a value for site_domain requires django.contrib.sites, which isn't installed"

CuriousLearner replied:
Yeah, looks good!

CuriousLearner commented:
Hey @adamalton 

I've fixed the docs. Separated the flags of management command from the arguments that `ping_google` function accepts. I've also added a release entry as well.

Please let me know if any more changes are needed.

CuriousLearner commented:
@carltongibson I think having it https by default would be good. Shall I go ahead and modify the PR, or should we wait for a discussion on the bug first?

carltongibson commented:
I'd say, _assume the best_, and make the change. (Release will need pulling into Breaking Changes.)

CuriousLearner commented:
@carltongibson Refactored this. I'm not sure how the breaking changes are mentioned in the release note. Is there any special syntax to follow, or anything to take care of?

CuriousLearner commented on tests/sitemaps_tests/test_management.py:
```diff
@@ -10,8 +10,16 @@ class PingGoogleTests(SitemapTestsBase):
         call_command('ping_google', 'foo.xml')
-        ping_google_func.assert_called_with(sitemap_url='foo.xml')
+        ping_google_func.assert_called_with(sitemap_url='foo.xml', is_http=False, site_domain=None)
+        call_command('ping_google', 'foo.xml', '-u')
```
Sure, that looks good.

CuriousLearner commented on docs/releases/2.2.txt:
```diff
@@ -355,6 +355,11 @@ Miscellaneous
+* ``ping_google`` management command now accepts two more optional arguments.
+  ``-d --site_domain`` indicates the site domain to be used if
+  ``django.contrib.sites`` is not used. ``-u --use-http`` indicates the
+  use of ``http`` instead of ``https`` scheme.
```
Yeah, makes sense. I'll edit this patch right away.

CuriousLearner commented:
@carltongibson All done here. 👍

CuriousLearner commented:
Thanks, @carltongibson & @adamchainz 🎉

carltongibson commented on AUTHORS:
```diff
@@ -747,6 +747,7 @@ answer newbie questions, and generally made Django that much better:
     Sam Newman <http://www.magpiebrain.com/>
     Sander Dijkhuis <sander.dijkhuis@gmail.com>
     Sanket Saurav <sanketsaurav@gmail.com>
+    Sanyam Khurana <sanyam.khurana01@gmail.com>
```
As per https://github.com/django/django/pull/10652/files#r234646818:

> Ooops. As Tim just pointed out on #10652, this isn't in the right place. (We only need it once 🙂)

CuriousLearner replied:
The one in this PR is sorted alphabetically. I will remove it from other PRs. Sounds good?

carltongibson replied:
Yeah, fine. 
Also can you [adjust the commit message](https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines) here too?

CuriousLearner replied:
@carltongibson I've reframed the commit message.

claudep commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "python.formatting.provider": "black"
+}
```
Probably not a wanted addition in Django?

CuriousLearner replied:
Sorry, I just recently switched to `vscode` and accidentally committed it. Fixing this!

carltongibson commented:
OK. Thanks for the input. De-RFC-ing.

carltongibson commented:
Hmmm. This is a bit of a pain. 😕 we can't provide the desired new usage from day one and a BC shim. Looks like we're saying "do this sub-optimal thing for ages, and we'll be breaking the old usage then". Part of me would rather just bite the bullet at the beginning. 

> ... if you're doing HTTPS properly... 

Yes, if. 🙂 I'd guess most **are** redirecting from http to https, even if they're not doing the rest. So, maybe `http` isn't that bad, but can we think of a decent way of handling the deprecation?

carltongibson commented on docs/ref/contrib/sitemaps.txt:
```diff
@@ -483,11 +483,24 @@ that: :func:`django.contrib.sitemaps.ping_google()`.
+        * ``site_domain``, If you are not using ``django.contrib.sites``, use this
+          to pass your site's domain (e.g., ``example.com``)
+
+        * ``is_http``, A boolean to indicate to use ``http://`` scheme instead of
```
Not sure about the inconsistency between `is_http` here and `--use_http` for the command option. (I think `use` for both.)

adamchainz replied:
agree

carltongibson commented on docs/ref/contrib/sitemaps.txt:
```diff
@@ -530,3 +543,13 @@ Once the sitemaps application is added to your project, you may also
+
+    .. versionadded:: 2.2
+
+    * ``-d --site_domain``, If you are not using ``django.contrib.sites``, use this
```
Do we prefer dashes to underscore in command options? (So `--site-domain` etc?)

KyeRussell replied:
Dashes. Definitely.

adamchainz replied:
+1

timgraham replied:
These should be documented using `.. django-admin-option::` as in `docs/ref/django-admin.txt` (and then linked from the release notes).

CuriousLearner replied:
I would also prefer dashes here, but I used underscore for consistency reasons. If you check the other argument, `sitemap_url` => https://github.com/django/django/pull/10651/files#diff-53e32f908b617b1896a4aacca34b4b25R9

CuriousLearner replied:
Hey @timgraham 

Sorry, but I'm confused about linking the option from release notes in `sitemaps.txt` documentation. Can you please help me here? I've addressed all other reviews.

carltongibson commented:
@MarkusH had [a suggestion for the deprecation on the mailing list](https://groups.google.com/d/topic/django-developers/4R5PNFnFWSw/discussion): 

> We could introduce a settings variable `SITEMAPS_PING_GOOGLE_HTTPS` that's part of newly created projects' settings (in https://github.com/django/django/blob/master/django/conf/project_template/project_name/settings.py-tpl) and set to `True`. In global_settings.py it defaults to `False`. Users could then turn it on for existing projects. Once the deprecation period is over we would simply remove the variable and thus enforce HTTPS. We could then also possibly issue a notice through the checks framework that the variable is now obsolete. 

@CuriousLearner Would you be able to add something along those lines this week (i.e. before the feature freeze)? It would need an addition to `docs/internals/deprecation.txt`.

charettes commented on docs/ref/contrib/sitemaps.txt:
```diff
@@ -530,3 +543,13 @@ Once the sitemaps application is added to your project, you may also
+    * ``-d --site_domain``, If you are not using ``django.contrib.sites``, use this
+        to pass your site's domain (e.g., ``example.com``)
+
+    * ``--use_http``, A boolean to indicate to use ``http://`` scheme instead of
```
Options usually use a `-` for separators.
```suggestion
    * ``--use-http``, A boolean to indicate to use ``http://`` scheme instead of
```

charettes commented on django/contrib/sitemaps/management/commands/ping_google.py:
```diff
@@ -7,6 +7,12 @@ class Command(BaseCommand):
     def add_arguments(self, parser):
         parser.add_argument('sitemap_url', nargs='?')
+        parser.add_argument('-d', '--site_domain', default=None)
+        parser.add_argument('--use_http', default=False, action='store_false')
```
Options usually use a `-` for separators.
```suggestion
        parser.add_argument('--use-http', default=False, action='store_false')
```

charettes commented on django/contrib/sitemaps/management/commands/ping_google.py:
```diff
@@ -7,6 +7,12 @@ class Command(BaseCommand):
 
     def add_arguments(self, parser):
         parser.add_argument('sitemap_url', nargs='?')
+        parser.add_argument('-d', '--site_domain', default=None)
```
Options usually use a `-` for separators.
```suggestion
        parser.add_argument('-d', '--site-domain', default=None)
```

timgraham replied:
I believe `default=None` isn't needed because that's the default (possibly the same with `default=False` below).

timgraham commented on django/contrib/sitemaps/management/commands/ping_google.py:
```diff
@@ -7,6 +7,12 @@ class Command(BaseCommand):
+        ping_google(
+            sitemap_url=options['sitemap_url'],
+            site_domain=options['site_domain'],
+            is_http=options['use_http']
```
Include a trailing comma in cases like this so that if more items are added later, this line doesn't have to be modified again.

timgraham commented on docs/ref/contrib/sitemaps.txt:
```diff
@@ -483,11 +483,24 @@ that: :func:`django.contrib.sitemaps.ping_google()`.
+
+        .. _`Google Webmaster Ping Tool`: https://www.google.com/webmasters/tools/ping
+
+        .. versionadded:: 2.2
```
This should appear below the options in a "self-contained" note as described in [Documenting new features](https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#documenting-new-features).

timgraham commented on docs/ref/contrib/sitemaps.txt:
```diff
@@ -483,11 +483,24 @@ that: :func:`django.contrib.sitemaps.ping_google()`.
-    sitemap by performing a reverse looking in your URLconf.
+    :func:`ping_google` takes the following optional arguments:
+
+        * ``sitemap_url`` which should be the absolute path to your site's sitemap (e.g.,
```
No need to indent for a list. Please be careful to wrap at 79 characters.

timgraham commented on docs/releases/2.2.txt:
```diff
@@ -355,6 +355,12 @@ Miscellaneous
   dashes, e.g. ``550e8400-e29b-41d4-a716-446655440000`` instead of
   ``550e8400e29b41d4a716446655440000``.
 
+* ``ping_google`` management command now accepts two more optional arguments.
```
"The ping_google management command now accepts the --site-domain and --use-http options. The first ... and the second ..."

That should be documented under "minor features" and only the backwards incompatible part about switch from http to https mentioned here.

carltongibson commented:
Hey @CuriousLearner. I think the small BC is OK. If you can make the review adjustments and uncheck `Patch needs improvement` on the ticket, that will be great. 

Thank you for your effort!

CuriousLearner commented:
Hi,

I've tried to address the reviews, however, I couldn't [wrap my head around linking back in the docs](https://github.com/django/django/pull/10651#discussion_r246593655) (not sure how do I do it, so would appreciate any help).

Thanks!

timgraham commented on docs/ref/contrib/sitemaps.txt:
```diff
@@ -530,3 +545,19 @@ Once the sitemaps application is added to your project, you may also
+
+.. versionadded:: 2.2
+
+Specifies the site domain to be used if ``django.contrib.sites`` is not used.
```
I don't understand this. You need to put django.contrib.sites in INSTALLED_APPS in order to use the ping_google command.

timgraham commented on docs/ref/contrib/sitemaps.txt:
```diff
@@ -530,3 +545,19 @@ Once the sitemaps application is added to your project, you may also
+
+.. versionadded:: 2.2
+
+Specifies the usage of ``http://`` scheme instead of ``https://``. Defaults to
```
In what URL?

timgraham commented on docs/releases/2.2.txt:
```diff
@@ -427,6 +430,9 @@ Miscellaneous
   dashes, e.g. ``550e8400-e29b-41d4-a716-446655440000`` instead of
   ``550e8400e29b41d4a716446655440000``.
 
+* The ``ping_google`` management command now uses ``https`` scheme by default.
```
For what URL?

timgraham commented on docs/ref/contrib/sitemaps.txt:
```diff
@@ -483,16 +483,31 @@ that: :func:`django.contrib.sitemaps.ping_google()`.
 
 .. function:: ping_google
```
The signature can be added here.

timgraham commented:
I'm working on a patch for just the "use http" changes so I can polish the docs at the same time.
